src/pages/instances/InstanceList.tsx

```diff
@@ -307,6 +344,65 @@ const InstanceList: FC = () => {
       };
     });
 
+  const profileSet = [
+    ...new Set(instances.flatMap((instance) => instance.profiles)),
+  ];
+
+  const searchAndFilterData: SearchAndFilterData[] = [
+    {
+      id: 1,
+      heading: "Status",
+      chips: instanceStatuses
+        .concat(
+          temporaryStatuses.filter((status) =>
+            instances.some((instance) => instance.status === status)
+          )
+        )
+        .map((status) => {
+          return { lead: "status", value: status };
+        }),
+    },
+    {
+      id: 2,
+      heading: "Instance type",
+      chips: instanceTypes.map((type) => {
+        return { lead: "type", value: type.uiName };
+      }),
+    },
+    {
+      id: 3,
+      heading: "Profile",
+      chips: profileSet.map((profile) => {
+        return { lead: "profile", value: profile };
+      }),
+    },
+  ];
+
+  const onSearchDataChange = (searchData: SearchAndFilterChip[]) => {
+    const newQueries = searchData
+      .filter((chip) => chip.quoteValue)
+      .map((chip) => chip.value);
+    const newStatuses = searchData
+      .filter((chip) => chip.lead === "status")
+      .map((chip) => chip.value as LxdInstanceStatus);
+    const newTypes = searchData
+      .filter((chip) => chip.lead === "type")
+      .map((chip) => {
+        return {
+          name: chip.value === "VM" ? "virtual-machine" : "container",
+          uiName: chip.value,
+        } as InstanceListType;
+      });
+    const newProfileQueries = searchData
+      .filter((chip) => chip.lead === "profile")
+      .map((chip) => chip.value);
+    if (!isEqual(queries, newQueries)) setQueries(newQueries);
```

Comment:
I think it is easier now to combine all states into a single state variable with a complex type:

```
interface InstanceFilter = {
queries: string[];
status: LxdInstanceStatus[];
...

const [filter, setFilter] = useState<InstanceFilter>()
```
Then we don't need the multple isEqual checks here, but just replace the whole object.

Reply:
Right, this way we can reduce the `isEqual` check to one instead of four, but you cannot replace the whole object blindly. You'd exceed the maximum update depth otherwise.